### PR TITLE
Fix safety issues with splitLines, lineSplitter, and indexOf

### PR DESCRIPTION
**Pull request by jmdavis**
This fixes it so that when types that implicitly convert to string are used with `splitLines`, `lineSplitter`, and `indexOf`, the implicit conversion is done at the call site so that we don't run into safety issues related to the conversion occurring inside the function (e.g. slicing a static array and thus returning a slice of local stack memory).

### CODE REVIEW

**Comment by MetaLang** [comment]
Is there a bugzilla entry for this specific safety issue?

**Comment by jmdavis** [comment]
> Is there a bugzilla entry for this specific safety issue?

I don't think so. Discussions on `isConvertibleToString` made it clear that it was inherently unsafe and needs to be removed, and these are just some of the changes that need to be made before we can do that.

**Comment by quickfur** [comment]
ping @MetaLang @jmdavis @andralex 

This has been rotting here for far too long. What needs to be done to get things going again?

**Comment by jmdavis** [comment]
Okay. I rebased it. As for "getting it going again," I don't know of anything that needs to be done on my end. It's just that it never really got more than a cursory look by anyone and then just sat here with no one doing anything about it. If my PRs related to `isConvertibleToString` would be merged, I could continue with working towards expunging it from Phobos, but I haven't bothered, because the existing PRs just sat here.

**Comment by quickfur** [comment]
Hmm. TBH I'm not a fan of introducing yet more overloads that technically don't really concern the user. Is there any way to make something like the following work?
````
ptrdiff_t indexOf(R)(R range, ...) {
    return indexOfImpl(range, ...);
}

ptrdiff_t indexOfImpl(C)(C[] arr, ...) if (isSomeChar!C) ..

ptrdiff_t indexOfImpl(R)(R range, ...) if (isInputRange!R && !isSomeChar!(ElementType!R)) ...
````

IOW, the user-facing API should be just a single overload that takes everything; what we do with it internally is none of the user's business.  (Though I vaguely recall there was some kind of problem with implicit conversions that would preclude this.)

**Comment by jmdavis** [comment]
> IOW, the user-facing API should be just a single overload that takes everything; what we do with it internally is none of the user's business. (Though I vaguely recall there was some kind of problem with implicit conversions that would preclude this.)

It is literally impossible to use one overload in this case. The _only_ way for the implicit conversions to work correctly is if they're done at the call site, which requires that the function take an array explicitly, and if the implicit conversion happens within the function, then you have an @safety problem (e.g. instead of slicing a static array at the call site, you slice a copy of it inside the function).

In order to fix all of the `isConvertibleToString` stuff, we have to have at least two overloads for each affected function. Yes, it sucks, but there's no other way to fix it. Really, most such functions should never have accepted implicitly convertible types in the first place, but many of them started by taking `string` or `const(char)[]` or whatnot, and unfortunately, static arrays implicitly convert to dynamic ones, so in order to avoid breaking code, you're forced to continue to accept the implicit conversions, and that means multiple overloads.

File: std/string.d [hunk 1]
```diff
@@ -357,10 +357,21 @@ alias CaseSensitive = Flag!"caseSensitive";
...
+}
+
+/// Ditto
+ptrdiff_t indexOf(C)(C[] s, dchar c, CaseSensitive cs = Yes.caseSensitive)
+if (isSomeChar!C)
+{
+    return _indexOf(s, c, cs);
+}
+
+ptrdiff_t _indexOf(Range)(Range s, dchar c, CaseSensitive cs = Yes.caseSensitive)
```
Comment (wilzbach) [inline std/string.d]: `private`

File: std/string.d [hunk 2]
```diff
@@ -485,10 +496,21 @@ if (isInputRange!Range && isSomeChar!(ElementEncodingType!Range) &&
...
+}
+
+/// Ditto
+ptrdiff_t indexOf(C)(C[] s, dchar c, size_t startIdx, CaseSensitive cs = Yes.caseSensitive)
+if (isSomeChar!C)
+{
+    return _indexOf(s, c, startIdx, cs);
+}
+
+ptrdiff_t _indexOf(Range)(Range s, dchar c, size_t startIdx, CaseSensitive cs = Yes.caseSensitive)
```
Comment (wilzbach) [inline std/string.d]: `private`

**Comment by jmdavis** [comment]
Okay. I made the helper functions private.

**Comment by quickfur** [comment]
@wilzbach Haha, you beat me to it again. :-P

File: std/string.d [hunk 3]
```diff
@@ -357,10 +357,21 @@ alias CaseSensitive = Flag!"caseSensitive";
...
+}
+
+/// Ditto
+ptrdiff_t indexOf(C)(C[] s, dchar c, CaseSensitive cs = Yes.caseSensitive)
+if (isSomeChar!C)
+{
+    return _indexOf(s, c, cs);
+}
+
+private ptrdiff_t _indexOf(Range)(Range s, dchar c, CaseSensitive cs = Yes.caseSensitive)
```
Comment (andralex) [inline std/string.d]: One future PR would do good to unify these overloads, put this implementation inside, etc...

Comment (wilzbach) [reply std/string.d]: That's not possible due to D's implicit conversion from static arrays to dynamic arrays. See @jmdavis's explanation here or https://github.com/dlang/phobos/pull/5410#discussion_r164238570

Comment (quickfur) [reply std/string.d]: Yeah, I'm of the opinion that we should deprecate implicitly converting static arrays to dynamic arrays (require explicit slicing, e.g., `staticArr[]` instead of just `staticArr`). That is an important first step towards making it possible to merge these overloads.

Comment (andralex) [reply std/string.d]: auto ref could help that situash

Comment (quickfur) [reply std/string.d]: @andralex: I think you misunderstood something. The problem is that `indexOf` was originally declared to accept arrays, something like this:
````
ptrdiff_t indexOf(C)(C[] str, dchar c, CaseSensitive cs = Yes.caseSensitive) ...
````
This original function accepted dynamic arrays, obviously, *but* due to static arrays implicitly converting to dynamic arrays, it also accepted static arrays as well:
````
string x; indexOf(x, 'a'); // OK
char[3] y; indexOf(y, 'b'); // also OK (implicit conversion of char[3] to char[])
````

Then we generalized it to take an input range instead:
````
ptrdiff_t indexOf(R)(R r, dchar c, CaseSensitive cs = Yes.caseSensitive)
if (isInputRange!R) ...
````

Unfortunately, static arrays are *not* input ranges (and rightly so, because you can't shrink their length so `popFront` cannot be implemented for them). So now we have the situation:
````
string x; indexOf(x, 'a'); // OK
char[3] y; indexOf(y, 'b'); // Compile error: breaks existing code
````
`auto ref` has nothing to do with this; the underlying problem is that the implicit conversion of static arrays to dynamic arrays only happens if the function parameter is explicitly declared as an array; if it's a generic type `R` with a sig constraint, this conversion never happens, so `R` winds up being `char[3]`, which is not an input range.  Adding `auto ref` *at most* only turns that into `ref char[3]`, which is *still* not an input range.

Furthermore, this is not fixable by shunting the array overload into a private overload set, e.g., I thought of doing this:
````
ptrdiff_t indexOf(R)(R r, dchar ch)
// accept everything here
{
    return _indexOfImpl(r, ch);
}

ptrdiff_t _indexOfImpl(R r, dchar ch)
if (isInputRange!R && !isSomeString!R) // exclude the char[3] case
{...}

ptrdiff_t _indexOfImpl(C)(C[] arr, dchar ch) // induce implicit conversion char[3] -> char[] here
{ ... }
````
Unfortunately, since static arrays are passed by value, this means that if the function returns a modified range, then `_indexOfImpl` would be slicing a local variable (function parameter of `indexOf`) instead of having the slice taken *in the caller's scope*. This would violate `@safe`.

Basically, it all boils down to static arrays being implicitly sliced into dynamic arrays causing problems with generalizing functions that take dynamic arrays to functions that take ranges. Not to mention, such implicit conversions are a bad idea to begin with, because you can inadvertently return an escaping reference to a local static array due to `T[n]` decaying into `T[]` autoomatically. Had we required explicit slicing on the caller's side, i.e., `indexOf(staticArr[])` instead of `indexOf(staticArr)`, none of this would have been a problem.

Comment (andralex) [reply std/string.d]: I wasn't misunderstanding. My point is very simple. The following:

```d
ptrdiff_t indexOf(R)(auto ref R r, dchar c, CaseSensitive cs = Yes.caseSensitive)
if (appropriately_chosen_constraints)
{
    static if (zis) { ... }
    else static if (zat) { ... }
    else static if (ze_ozer) { ... }
}
```

covers everything: input ranges, dynamic arrays, static arrays (efficiently, too) with no problem. The only trick is defining the constraint properly.

Comment (jmdavis) [reply std/string.d]: In this particular case, yes, that could be done. Because no reference to the string is returned, but in any of the cases where the string does escape, the implicit conversion must be done at the call site. Regardless, I think that we need to expunge `isConvertibleToString`, because it makes it too easy to get this wrong and can't be used safely in any situation where a reference to the resultant string escapes. Many of the places that it's currently being used in Phobos are not `@safe` when used with something like a static array.
